File: v24_interpreter/java/SimpleInterpreter3.java
package v24_interpreter.java;

// AbstractExpression
interface Expression {
    void interpret();
}

// TerminalExpression for PRINT
class PrintExpression implements Expression {
    private String message;

    public PrintExpression(String message) {
        this.message = message;
    }

    @Override
    public void interpret() {
        System.out.println(message);
    }
}

// NonTerminalExpression for REPEAT
class RepeatExpression implements Expression {
    private int repeatCount;
    private Expression expression;

    public RepeatExpression(int repeatCount, Expression expression) {
        this.repeatCount = repeatCount;
        this.expression = expression;
    }

    @Override
    public void interpret() {
        for (int i = 0; i < repeatCount; i++) {
            expression.interpret();
        }
    }
}

// TerminalExpression for REVERSE
class ReverseExpression implements Expression {
    private String message;

    public ReverseExpression(String message) {
        this.message = message;
    }

    @Override
    public void interpret() {
        System.out.println(new StringBuilder(message).reverse().toString());
    }
}

// Client class
public class SimpleInterpreter3 {
    public static void main(String[] args) {
        String command = "REPEAT 3 TIMES: REVERSE Hello";

        // Split the command into words based on whitespace
        String[] words = command.split(" ");

        //Handle the command
        if (words[0].equalsIgnoreCase("REPEAT")) {
            int repeatCount = Integer.parseInt(words[1]);

            // Create the TerminalExpression for the desired command
            Expression expression;
            if (words[3].equalsIgnoreCase("PRINT")) {
                expression = new PrintExpression(words[4]);
            } else if (words[3].equalsIgnoreCase("REVERSE")) {
                expression = new ReverseExpression(words[4]);
            } else {
                throw new IllegalArgumentException("Unsupported command: " + words[3]);
            }

            // Create the NonTerminalExpression for REPEAT
            Expression repeatExpression = new RepeatExpression(repeatCount, expression);

            // Interpret the command
            repeatExpression.interpret();
        }
    }
}

File: v24_interpreter/python/simple_interpreter3.py
class Expression:
    def interpret(self):
        pass

class PrintExpression(Expression):
    def __init__(self, message):
        self.message = message

    def interpret(self):
        print(self.message)

class RepeatExpression(Expression):
    def __init__(self, repeat_count, expression):
        self.repeat_count = repeat_count
        self.expression = expression

    def interpret(self):
        for i in range(self.repeat_count):
            self.expression.interpret()

class ReverseExpression(Expression):
    def __init__(self, message):
        self.message = message

    def interpret(self):
        print(self.message[::-1])

def main():
    command = "REPEAT 3 TIMES: REVERSE Hello"

    # Split the command into words based on whitespace
    words = command.split(" ")

    # Handle the command
    if words[0].upper() == "REPEAT":
        repeat_count = int(words[1])

        # Create the TerminalExpression for the desired command
        if words[3].upper() == "PRINT":
            expression = PrintExpression(words[4])
        elif words[3].upper() == "REVERSE":
            expression = ReverseExpression(words[4])
        else:
            raise ValueError(f"Unsupported command: {words[3]}")

        # Create the NonTerminalExpression for REPEAT
        repeat_expression = RepeatExpression(repeat_count, expression)

        # Interpret the command
        repeat_expression.interpret()

if __name__ == "__main__":
    main()


